external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala

```diff
@@ -0,0 +1,174 @@
+    .expireAfterAccess(cacheExpireTimeout, TimeUnit.MILLISECONDS)
+    .removalListener(removalListener)
+    .build[String, Producer]()
+
+  ShutdownHookManager.addShutdownHook { () =>
+    clear()
+  }
+
+  private def createKafkaProducer(
+    producerConfiguration: ju.Map[String, Object]): Producer = {
+    val uid = getUniqueId(producerConfiguration)
+    val kafkaProducer: Producer = new Producer(producerConfiguration)
+    guavaCache.put(uid.toString, kafkaProducer)
+    logDebug(s"Created a new instance of KafkaProducer for $producerConfiguration.")
+    kafkaProducer
+  }
+
+  private def getUniqueId(kafkaParams: ju.Map[String, Object]): String = {
+    val uid = kafkaParams.get(CanonicalizeKafkaParams.sparkKafkaParamsUniqueId)
+    assert(uid != null, s"KafkaParams($kafkaParams) not canonicalized.")
+    uid.toString
+  }
+
+  /**
+   * Get a cached KafkaProducer for a given configuration. If matching KafkaProducer doesn't
+   * exist, a new KafkaProducer will be created. KafkaProducer is thread safe, it is best to keep
+   * one instance per specified kafkaParams.
+   */
+  private[kafka010] def getOrCreate(kafkaParams: ju.Map[String, Object]): Producer = synchronized {
+    val params = if (!CanonicalizeKafkaParams.isCanonicalized(kafkaParams)) {
+      CanonicalizeKafkaParams.computeUniqueCanonicalForm(kafkaParams)
+    } else {
+      kafkaParams
+    }
+    val uid = getUniqueId(params)
+    Option(guavaCache.getIfPresent(uid)).getOrElse(createKafkaProducer(params))
+  }
+
+  /** For explicitly closing kafka producer */
+  private[kafka010] def close(kafkaParams: ju.Map[String, Object]): Unit = {
+    val params = if (!CanonicalizeKafkaParams.isCanonicalized(kafkaParams)) {
+      CanonicalizeKafkaParams.computeUniqueCanonicalForm(kafkaParams)
+    } else kafkaParams
+    val uid = getUniqueId(params)
+    guavaCache.invalidate(uid)
+  }
+
+  /** Auto close on cache evict */
+  private def close(uid: String, producer: Producer): Unit = {
+    try {
+      val outcome = CanonicalizeKafkaParams.remove(
+        new ju.HashMap[String, Object](
+          Map(CanonicalizeKafkaParams.sparkKafkaParamsUniqueId -> uid).asJava))
+      logDebug(s"Removed kafka params from cache: $outcome.")
+      logInfo(s"Closing the KafkaProducer with uid: $uid.")
+      producer.close()
+    } catch {
+      case NonFatal(e) => logWarning("Error while closing kafka producer.", e)
+    }
+  }
+
+  private def clear(): Unit = {
+    logInfo("Cleaning up guava cache.")
+    guavaCache.invalidateAll()
+  }
+
+  // Intended for testing purpose only.
+  private def getAsMap: ConcurrentMap[String, Producer] = guavaCache.asMap()
+}
+
+/**
+ * Canonicalize kafka params i.e. append a unique internal id to kafka params, if it already does
+ * not exist. This is done to ensure, we have only one set of kafka parameters associated with a
+ * unique ID.
+ */
+private[kafka010] object CanonicalizeKafkaParams extends Logging {
```

Comment:
This seems kind of complicated also.  Since we know these are always coming from `Data[Stream/Frame]Writer` and that will always give you `Map[String, String]` (and we expect the number of options here to be small).  Could we just make the key for the cache a sorted `Seq[(String, String)]` rather than invent another GUID?